### ISSUE TITLE
orbital SQL produces wrong prediction for output_probability.0 for XGBoostClassifier (wrong margin transformation)

### ISSUE THREAD

**Issue by emilyriederer** [posit-dev/orbital#53]
`orbital` SQL produces the wrong predictions for non-events when using an `XGBoostClassifier` in my pipeline because it is applying the post-processing in the wrong order (`post_transform(1 - base_margin)` instead of `1 - post_transform(base_margin)`) when converting GBM base margin into a probability score. 

Specifically, I believe the problem is that [this happens](https://github.com/posit-dev/orbital/blob/08302c9b1c403d209d00d7c5974b2fd17b51919b/src/orbital/translation/steps/trees/classifier.py#L166) before [this happens](https://github.com/posit-dev/orbital/blob/08302c9b1c403d209d00d7c5974b2fd17b51919b/src/orbital/translation/steps/linearclass.py#L99). 

----

Reproducible example below (`orbital` 0.2.2`, Python 3.12): 

```
import orbital
import duckdb
import sqlglot
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_classification

# mock data
X,y = make_classification()
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# mock pipeline
pipeline = Pipeline(
    [
        ("preprocess", ColumnTransformer([("scaler", StandardScaler(with_std=False), [0,1,2])],
                                        remainder="passthrough")),
        ("gbm", GradientBoostingClassifier(max_depth = 1, n_estimators = 1)),
    ]
)
pipeline.fit(X_train, y_train)

# render SQL from orbital
n_cols = len(X_train[0])
nm_cols = [f"var_{i}" for i in range(n_cols)]
feat_dict = {}
for n in nm_cols:
    feat_dict[n] = orbital.types.DoubleColumnType()
orbital_pipeline = orbital.parse_pipeline(pipeline, features=feat_dict)
sql_raw = orbital.export_sql("DATA_TABLE", orbital_pipeline, dialect="duckdb")

# score model
DATA_TABLE = pd.DataFrame(X_train, columns = nm_cols)
duckdb.sql(sql_raw).df().head(5)
```

This produces the following output which is clearly wrong since the two probability scores do not add to 1. 

<img width="661" height="250" alt="Image" src="https://github.com/user-attachments/assets/fca6d20c-2474-4f5f-b51e-096188a0a96a" />

If we compare to predicted probability from python, we can see that the event probability matches but the non-event probability does not: 

```
pipeline.predict_proba(X_train)[:5]
``` 

```
array([[0.53677557, 0.46322443],
       [0.46671596, 0.53328404],
       [0.53677557, 0.46322443],
       [0.46671596, 0.53328404],
       [0.46671596, 0.53328404]])
```

I looked at the SQL code to see what could be going on by writing it to a file:

```
import sqlglot

with open("model_test.sql", "w") as file:
    sql_fmt = sqlglot.transpile(sql_raw, write="duckdb", identify=True, pretty=True)[0]
    file.write(sql_fmt)
```

Here is the output:

```
SELECT
  CAST(CASE
    WHEN CASE
      WHEN "t0"."var_10" <= -0.15715152025222778
      THEN -0.1473684161901474
      ELSE 0.13333334028720856
    END > 0.5
    THEN 1
    ELSE 0
  END AS BIGINT) AS "output_label",
  1 / (
    EXP(
      -(
        1.0 - CASE
          WHEN "t0"."var_10" <= -0.15715152025222778
          THEN -0.1473684161901474
          ELSE 0.13333334028720856
        END
      )
    ) + 1
  ) AS "output_probability.0",
  1 / (
    EXP(
      -CASE
        WHEN "t0"."var_10" <= -0.15715152025222778
        THEN -0.1473684161901474
        ELSE 0.13333334028720856
      END
    ) + 1
  ) AS "output_probability.1"
FROM "DATA_TABLE" AS "t0"
```

We can see that for `output_probability.1` it is correctly applying the sigmoid transformation ( `1/( e^-z + 1)` which is equivalent to `e^z  / (e^z + 1)`) . However, for `output_probability.0` it is doing `1 / (e^-(1-z) + 1)` instead of something equivalent to `1 / (e^z + 1)` which would give the desired result.  

Needless to say, if I swap in a `RandomForestClassifier`, this works as expected since no post-transformations are required.

**Comment by amol-** [comment]
@emilyriederer can you confirm this only happens in the case of binary classification and multi-class classification instead works as expected?

**Comment by emilyriederer** [comment]
@amol- That is correct! Only happens in binary case. Here's a multiclass version that works as intended.

I believe only the binary uses the `1 - prediction` [logic](https://github.com/posit-dev/orbital/blob/08302c9b1c403d209d00d7c5974b2fd17b51919b/src/orbital/translation/steps/trees/classifier.py#L166) which seems to be the root cause. 

```
import orbital
import duckdb
import sqlglot
import pandas as pd
import numpy as np
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_classification

# mock data
X_train, y_train = make_classification(random_state=504, n_classes=3, n_informative=5)
X_train = X_train.round(3)

# mock pipeline
pipeline = Pipeline([
    ("prep", ColumnTransformer([("scaler", StandardScaler(), [])], remainder="passthrough")),
    ("gbm", GradientBoostingClassifier(max_depth = 1, n_estimators = 1, random_state=102)),
])
pipeline.fit(X_train, y_train)

# render SQL from orbital
n_cols = len(X_train[0])
nm_cols = [f"var_{i}" for i in range(n_cols)]
feat_dict = {}
for n in nm_cols:
    feat_dict[n] = orbital.types.DoubleColumnType()
orbital_pipeline = orbital.parse_pipeline(pipeline, features=feat_dict)
sql_raw = orbital.export_sql("DATA_TABLE", orbital_pipeline, dialect="duckdb")

# structure of table
DATA_TABLE = pd.DataFrame(X_train, columns = nm_cols)
df_preds = duckdb.sql(sql_raw).df()
df_preds = df_preds.assign(
    p = lambda d: d['output_probability.0'] + 
                  d['output_probability.1'] + 
                  d['output_probability.2']
                  )

# check sum 
np.all(np.isclose(df_preds['p'],1))
```
